README.md
# An extension for c coders to create comments quickly
## 1. configuation
Once you install the chelper into the VSCODE,  you have to do the configuration 
for the chelper.
```
{
    "chelper.user_email": "your email address",
    "chelper.user_name": "your name ",
    "chelper.user_company": "your company"
    "chelper.user_copyrights": "All rights reserved."
}
```
## 2. base command 
1. type the keywords and then press CTRL+1 to get the comment 
2. the keywords lists bellow:
- hi
> to add a modify history
- ad/dd/md
> to add a block comment
> ad means add , dd means delete, md means modify
- func
> add a function comment
- hfile
> add a new header file
- cfile
> add a new c source file
- head
> add a file header


CHANGELOG.md
# Change Log
All notable changes to the "chelper" extension will be documented in this file.

## [0.0.3]
- fix feature 'func' issue
    - 1.when the text lines that from the position of the cursor to the end of the document are less than 16, the extension can't work correctly.
         So when execute 'func' function, we compute the left lines count to make feature right.
    - 2.when there is just one 'void' parament for the function, the extension can't get the parament correctly.

## [0.0.2]
- modify the extension description string and README.md 

## [0.0.1]
- Initial release

baseLib.js
function getTimeStr() {
    var ds = new Date();
    return ds.getFullYear().toString() + "/" + (ds.getMonth() +1).toString() + "/" 
    + ds.getDate().toString() + " " + ds.getHours().toString() + ":" + ds.getMinutes().toString() + ":" + ds.getSeconds().toString();

}


function getFileNamePart(filename_with_type) {
    var strArray = filename_with_type.split('.');
    if (null != strArray)
        return strArray[0];
    
    return filename_with_type;

}

function alignStr2Width(str, width, fill) {
    var count = 0;
    var oldLen = str.length;
    if (oldLen>= width)
        return str;
    
    for (count=0; count<(width-oldLen); count++) {
        str += fill;
    }
    return str;

}
module.exports.getTimeStr = getTimeStr;
module.exports.getFileNamePart = getFileNamePart;
module.exports.alignStr2Width = alignStr2Width;


cDygCmt.js

var baseLib = require('./baseLib');
var util = require('util');

function getBlockCmt(comment) {
        var blockStart = "/**\n";
        var blockEnd = "*/\n";
        return blockStart + comment + blockEnd;

    }

    function getBlockCmtCxt(user, time, brief) {
        var blockComCmt = "\
* @brief  %s\n*\n*\n*\n\
* @author %s\n\
* @email  %s\n\
* @date   %s\n\
";
        return util.format(blockComCmt, brief, user._name, user._email, time);
    }


    function getCopyRights(user, other) {
        return util.format("* CopyRights(C) %s %s \n*%s\n", user._company, user._copyrights, other);

    }

    function getFileHead(user, time, brief, file) {
        var crs = getCopyRights(user, brief);
        var comTxt = getBlockCmtCxt(user, time , file)
        var comment = getBlockCmt(crs + comTxt);        

        return comment;  
    }
    function getCHeaderTemp(file) {

        var hDefine =  baseLib.getFileNamePart(file).toUpperCase();
        var cHeaderFormat = "#ifndef __%s_H__\n\
#define __%s_H__\n\
#ifdef __cplusplus\n\
extern \"C\" {\n\
#endif\n\
/*----------------------------------------------*\n\
* file including instruction                   *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
* declare extern variaries                     *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
* declare extern functions                     *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
*declare inner functions                             *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
* global variaries                              *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
* module variaries                              *\n\
*----------------------------------------------*/\n\n\n\
/*----------------------------------------------*\n\
* constants                                      *\n\
*----------------------------------------------*/\n\n\n\n\
/*----------------------------------------------*\n\
* marcro                                       *\n\
*----------------------------------------------*/\n\n\n\
#ifdef __cplusplus\n\
}\n\
#endif\n\
#endif /* __%s_H__ */\n";           
        return util.format(cHeaderFormat, hDefine, hDefine, hDefine);
    }

    function getOneParam(ptype, pname) {
        var funcParamFmt = "* @param  %s %s\n";
        return util.format(funcParamFmt, baseLib.alignStr2Width(ptype, 16, " "), baseLib.alignStr2Width(pname, 16, " "));
    }

    function getReturn(preturn) {
        var returnFmt = "* @return %s \n";
        return util.format(returnFmt, preturn);
    }


module.exports.getReturn = getReturn;
module.exports.getOneParam = getOneParam;
module.exports.getBlockCmt = getBlockCmt;
module.exports.getBlockCmtCxt = getBlockCmtCxt;
module.exports.getCopyRights =getCopyRights;
module.exports.getFileHead = getFileHead;
module.exports.getCHeaderTemp = getCHeaderTemp;


cHelper.js

var baseLib = require('./baseLib');
var userInfo = require('./userInfo');
var hc = require('./cHelper_handle');
var fc = require('./cHelper_format');



class baseCmd{
    constructor(key_str, des, handle, format) {
        this._key_str = key_str;
        this._cmd_des = des;
        this._handle = handle;
        this._format = format;        
    }

}
class mainProcess {
    constructor() {
        this._cmdHash = {};
        this._cmdHash["hfile"] = new baseCmd("hfile", "add a new head file", hc.handleCollections.comm, fc.formatCollections.headFile);
        this._cmdHash["cfile"] = new baseCmd("cfile", "add a new c file", hc.handleCollections.comm, fc.formatCollections.sourceFile);
        this._cmdHash["hi"] = new baseCmd("hi", "modify list ", hc.handleCollections.comm, fc.formatCollections.history);
        this._cmdHash["ad"] = new baseCmd("ad", "the comment for adding", hc.handleCollections.comm, fc.formatCollections.block);
        this._cmdHash["md"] = new baseCmd("md", "the comment for modifing", hc.handleCollections.comm, fc.formatCollections.block);                       
        this._cmdHash["dd"] = new baseCmd("dd", "the comment for deleting", hc.handleCollections.comm, fc.formatCollections.block);
        this._cmdHash["func"] = new baseCmd("func", "add a new function or function comment", hc.handleCollections.comm, fc.formatCollections.func);
        this._cmdHash["head"] = new baseCmd("head", "add a new file head", hc.handleCollections.comm, fc.formatCollections.fileHead);
    }

    run(vscode) {
            var textEd = vscode.window.activeTextEditor;
            var sel = textEd.selection;
            var time = baseLib.getTimeStr();
            var user = new userInfo.user(vscode); 
            var str = textEd.document.lineAt(sel.start.line);
            var key = str.text.trim();
            if (null == sel) 
            {
                return;
            }
        
            if (this._cmdHash[key] == null)
            {
                return;
            }

            try
            {
                var entry = this._cmdHash[key];
                entry._handle(entry, vscode, sel, time, user);
            }
            catch(err){
                console.log("cHelper exception: " + err.name + " message: " + err.message);
            }
        }
}

module.exports.mainProcess=mainProcess;

ideComm.js

    function openNewDocAndFill(vscode, uri, context) {
        var nPos = new vscode.Position(1, 1);
        vscode.workspace.openTextDocument(uri).then(doc => {
            vscode.window.showTextDocument(doc, 0, true).then (editor =>
                editor.edit(function (eb) 
                {
                    eb.insert(nPos, context);
            }));
        });

    }

    function   insertText2Doc(vscode, sel, str) {
                var nStart = new vscode.Position(sel.start.line, 0);
                var end = new vscode.Position(sel.end.line, sel.end.character);
                var nSel = new vscode.Selection(nStart, end);

                vscode.window.activeTextEditor.edit(function (eb) 
                {
                    eb.replace(nSel, "");
                    eb.replace(nStart, str);
                });
            }

    function getMoreLinesFromDoc(vscode, sel, lines) {
        var doc =  vscode.window.activeTextEditor.document;
        var count = sel.end.line;
        var tmp_str = "";
        var parseStr = "";
        var maxCnt = 0;

        // 2017-02-07, yuch, add a condition to set the correct left count as the limitation
        if (doc.lineCount <= (sel.end.line+lines))
            maxCnt = doc.lineCount;
        else
            maxCnt = sel.end.line + lines;
        
        while(++count < maxCnt) {

            tmp_str = doc.lineAt(count).text;
            if (0  == tmp_str.trim().length)
            {
                continue;
            }
            parseStr += " " + tmp_str;
        }
        return parseStr;
    }




module.exports.openNewDocAndFill=openNewDocAndFill;
module.exports.insertText2Doc=insertText2Doc;
module.exports.getMoreLinesFromDoc=getMoreLinesFromDoc;

extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
var vscode = require('vscode');
var chelper = require('./cHelper');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {

    // Use the console to output  diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "chelper" is now active!');
    var helper = new chelper.mainProcess();
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    var disposable = vscode.commands.registerCommand('extension.chelper', function () {
        // The code you place here will be executed every time your command is executed
        var win =  vscode.window.activeTextEditor;
        if (null == win)
        {
            vscode.window.showInformationMessage("can't get valiable handle for the window");
            return null;
        }
        helper.run(vscode);
    });

    context.subscriptions.push(disposable);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;

cHelper_format.js

var path = require('path');
var util = require('util');
var ideComm = require('./ideComm');
var cDygCmt = require('./cDygCmt');


class formatCollections {
    static block(cmd, vscode, sel, time, user) 
    {
        var help_str = {"dd":"delete", "ad":"add", "md":"modify"};
        var flag = 1; 
        var titleStr = util.format("%s %s by %s", time.toString(), help_str[cmd._key_str], user._name);
 
        var format = "/* %s  Begin*/\n#if %d \n\n#endif\n/* %s  End*/\n";
        if (cmd._key_str == "dd")
            flag   = 0;
        return util.format(format, titleStr, flag, titleStr);
    }


    static fileHead(cmd, vscode, sel, time, user) 
    {
        var filePath = path.parse(vscode.window.activeTextEditor.document.uri.fsPath);
        return cDygCmt.getFileHead(user, time, "", filePath.name+filePath.ext);
    }
    
    static history(cmd, vscode, sel, time, user)  
    {
        var format = "%s %s description: ";

        return util.format(format, time, user._name);
    }

    static headFile(cmd, vscode, sel, time, user) {
        vscode.window.showInputBox({prompt:"new c head file", placeHolder:"file name"}).then(fileName  => {
            var fileHead = cDygCmt.getFileHead(user, time, "", fileName);
            var cHead = cDygCmt.getCHeaderTemp(fileName);
            var comment = fileHead + cHead;
            var tUri = vscode.Uri.parse("untitled:" +  path.join(vscode.workspace.rootPath, fileName));
                ideComm.openNewDocAndFill(vscode, tUri, comment);
        });            
  

        return "";
    }

    static sourceFile(cmd, vscode, sel, time, user) 
    {
        vscode.window.showInputBox({prompt:"new c source file", placeHolder:"file name"}).then(fileName  => {
            var tUri = vscode.Uri.parse("untitled:" +  path.join(vscode.workspace.rootPath, fileName));
                var comment = cDygCmt.getFileHead(user, time, "", fileName);
                ideComm.openNewDocAndFill(vscode, tUri, comment);

            });
    }



    static func(cmd, vscode, sel, time, user) 
    {
        var commHead = cDygCmt.getBlockCmtCxt(user, time, "");
        var returnStr = "";
        var textCont = "";
        var blockCmtReg = /\/\*.*?\*\//g;
        var funcReg = /((?:static|inline|extern)?\s*?(?:const)?\s*(?:unsigned|struct)?\s*?\w+\s*?\*{0,2})\s*([-_\w]+)\s*['(']([^'(']*)[')']\s*\{/;
        //2017-02-07, yuch, modify the regx to support void parament
        var paramReg = /\s*((?:const)?\s*(?:unsigned|struct)?\s*[-|_|\w]+\s*\*{0,2})\s*([-|_|\w]*)\s*,?/g;
        var resArray = null;
        var paramStr = "";
        var result = null;

        //get 16 non-empty lines
        textCont = ideComm.getMoreLinesFromDoc(vscode, sel, 16);

        textCont = textCont.replace(blockCmtReg, "");

        //get func implementation string
        result = funcReg.exec(textCont);
        if (result == null)
            return;
        returnStr = cDygCmt.getReturn(result[1]);
        commHead = cDygCmt.getBlockCmtCxt(user, time, result[2]);
        while ((resArray = paramReg.exec(result[3])) != null)
        {
            paramStr += cDygCmt.getOneParam(resArray[1], resArray[2]);

        }
        return cDygCmt.getBlockCmt(commHead + paramStr + returnStr);
    }


}

module.exports.formatCollections=formatCollections;